File: src/modules/user/services/user-update.service.ts
import { hash } from 'bcryptjs';
import { UserUpdateDto } from '../domain/dto/user-update.dto';
import { UserEntity } from '../domain/entities/user.entity';
import { UserRequest } from '../../../shared/types/user-request';
import { prisma } from '../../../infra/database/prisma.service';
import {
  ConflictException,
  NotFoundException,
} from '../../../shared/exceptions';
import { userRoleValidator } from '../../../shared/validators/user-role.validator';

class UserUpdateService {
  async execute(
    id: number,
    dto: UserUpdateDto,
    userRequest: UserRequest,
  ): Promise<UserEntity> {
    const user = await prisma.user.findFirst({
      where: {
        id: id,
        organizationId: userRequest?.organizationId ?? undefined,
        companyId: userRequest?.companyId ?? undefined,
        deletedAt: null,
      },
    });

    if (!user) throw new NotFoundException('Usuário não encontrado!');

    const existingUser = await prisma.user.findFirst({
      where: {
        NOT: {
          id: id,
        },
        OR: [
          {
            name: dto.name,
          },
          {
            email: dto.email,
          },
        ],
        deletedAt: null,
      },
    });

    if (existingUser) throw new ConflictException('Usuário já cadastrado!');

    if (dto.roleId) await userRoleValidator(dto?.roleId, userRequest);

    const updatedUser = await prisma.user.update({
      data: {
        name: dto.name,
        email: dto.email,
        password: await hash(dto.password, 10),
        gender: dto.gender,
        phone: dto.phone,
        role: dto.roleId
          ? {
              connect: {
                id: dto?.roleId,
              },
            }
          : undefined,
        status: dto.status,
      },
      where: {
        id: id,
      },
      include: {
        role: true,
        organization: true,
        company: true,
        avatar: true,
      },
    });

    return new UserEntity(updatedUser);
  }
}

export const userUpdateService = new UserUpdateService();

File: src/modules/user/services/user_update_service.py
from django.db.models import Q
from src.shared.types.user_request import UserRequest
from src.modules.role.models import Role
from src.modules.user.models import User
from src.modules.user.domain.dto.user_update_dto import (
    UserUpdateDTO,
)
from src.shared.exceptions.conflict_exception import ConflictException
from src.shared.exceptions.not_found_exception import NotFoundException
from src.shared.validators.user_role_validator import (
    user_role_validator,
)
from src.shared.utils.hash import hash
from src.modules.user.validators.user_password_validator import user_password_validator


class UserUpdateService:
    @classmethod
    def execute(self, id: int, dto: UserUpdateDTO, user_request: UserRequest):

        user = User.objects.filter(
            Q(id=id, deleted_at=None)
            & (
                Q(organization_id=user_request.organization_id)
                if user_request and user_request.organization_id
                else Q()
            )
            & (
                Q(company_id=user_request.company_id)
                if user_request and user_request.company_id
                else Q()
            )
        ).first()

        if user is None:
            raise NotFoundException("Usuário não encontrado!")

        existing_user = User.objects.filter(
            ~Q(id=id) & (Q(name=dto.name) | Q(email=dto.email)) & Q(deleted_at=None)
        ).first()

        if existing_user is not None:
            raise ConflictException("Usuário já cadastrado!")

        if dto.role_id is not None:
            user_role_validator(dto.role_id, user_request)

        user_password_validator(dto.password)

        if dto.name is not None:
            user.name = dto.name
        if dto.email is not None:
            user.email = dto.email
        if dto.password is not None:
            user.password = hash(dto.password)
        if dto.phone is not None:
            user.phone = dto.phone
        if dto.status is not None:
            user.status = dto.status
        if dto.role_id is not None:
            user.role = Role.objects.get(id=dto.role_id)

        user.save()

        return user
